### PR TITLE
Consolidate root/utils tests

### PR DESCRIPTION
**Pull request by detiber**
- Consolidate tests between the root of the repo and utils

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -4,17 +4,34 @@ sudo: false
...
   - "2.7"
   - "3.5"
 
 install:
-  - pip install -r requirements.txt
   - pip install tox-travis
 
 script:
-  # TODO(rhcarvalho): check syntax of other important entrypoint playbooks
-  - ansible-playbook --syntax-check playbooks/byo/config.yml
```
Comment (detiber) [inline .travis.yml]: syntax check moved into tox

File: .travis.yml [hunk 2]
```diff
@@ -4,17 +4,34 @@ sudo: false
...
+  irc:
+    channels:
+      - chat.freenode.net#openshift-dev
+    on_success: change
+    on_failure: always
+    template:
+      - "%{repository}#%{build_number} (%{branch} - %{commit} : %{author}): %{message}"
+      - "Change view : %{compare_url}"
+      - "Build details : %{build_url}"
+      - "sdodson jdetiber: ^"
```
Comment (detiber) [inline .travis.yml]: notifications are for giving us an overview when the build is broken post-merge, since we've had some issues with generate_validation being broken after a merge.

File: .flake8 [hunk 3]
```diff
@@ -1,4 +1,5 @@
 [flake8]
-exclude=.tox,utils,inventory
+# TODO: cleanup flake8 issues with utils/test/*
```
Comment (rhcarvalho) [inline .flake8]: To be done in this PR?

File: requirements.txt [hunk 4]
```diff
@@ -1,5 +1,6 @@
 ansible>=2.2
-six
+click
 pyOpenSSL
-PyYAML
-ruamel.yaml
+# We need to disable ruamel.yaml for now because of test failures
+#ruamel.yaml
```
Comment (detiber) [inline requirements.txt]: Test fixes that work for both PyYAML and ruamel.yaml will be handled in a followup PR.

File: pytest.ini [hunk 5]
```diff
@@ -4,8 +4,6 @@ norecursedirs =
     __pycache__
     cover
     docs
-    # utils has it's own config
-    utils
```
Comment (rhcarvalho) [inline pytest.ini]: :heart:

File: .travis.yml [hunk 6]
```diff
@@ -4,17 +4,34 @@ sudo: false
...
-  - pip install -r requirements.txt
   - pip install tox-travis
 
 script:
-  # TODO(rhcarvalho): check syntax of other important entrypoint playbooks
-  - ansible-playbook --syntax-check playbooks/byo/config.yml
   - tox
-  - cd utils && tox
+
+notifications:
```
Comment (rhcarvalho) [inline .travis.yml]: Is this indeed intended or was just part of debugging coveralls?
I'm ok with it, just double checking ;)

File: tox.ini [hunk 7]
```diff
@@ -1,21 +1,24 @@
...
 minversion=2.3.1
 envlist =
-    py{27,35}-ansible22-{pylint,unit,flake8,yamllint,generate_validation}
+    py{27,35}-{flake8,pylint,unit}
+    py27-{yamllint,ansible_syntax,generate_validation}
 skipsdist=True
 skip_missing_interpreters=True
 
 [testenv]
+skip_install=True
```
Comment (detiber) [inline tox.ini]: No need to install for the root setup.py

File: requirements.txt [hunk 8]
```diff
@@ -1,5 +1,6 @@
 ansible>=2.2
-six
+click
 pyOpenSSL
-PyYAML
-ruamel.yaml
+# We need to disable ruamel.yaml for now because of test failures
```
Comment (rhcarvalho) [inline requirements.txt]: Hmm, the test fix didn't merge yet, right?
I got the error today when my venv had ruamel.yaml installed.

File: tox.ini [hunk 9]
```diff
@@ -1,21 +1,24 @@
...
 deps =
+    -rrequirements.txt
```
Comment (detiber) [inline tox.ini]: Previously the packages in requirements.txt were pulled in through the install process, which was not allowing us to take advantage of cached packages.

File: .flake8 (hunk 3 above)
Comment (detiber) [inline .flake8]: No, didn't want to add any more scope creep. Figured it was better to handle as we refactor the tests to take advantage of pytest.

File: requirements.txt (hunk 8 above)
Comment (detiber) [inline requirements.txt]: Not yet, I was going to do that as a followup PR

File: .travis.yml (hunk 6 above)
Comment (detiber) [inline .travis.yml]: Unrelated to coveralls, but seemed innocuous to add here since I was already updating the travis config.

Adding the notifications to help with the generate_validation errors we are seeing post-merge

**Comment by detiber** [comment]
aos-ci-test

**Comment by detiber** [comment]
@sdodson improved notifications for travis builds post-merge are included here.

File: .flake8 (hunk 3 above)
Comment (rhcarvalho) [inline .flake8]: Cool. I may have this covered. I'm cleaning up all tests to be more pytest-ish.
Got it it to a working state but needs some clean up -- tomorrow, as it is quite late here.

#3571

File: .coveragerc [hunk 10]
```diff
@@ -4,16 +4,17 @@ omit =
...
     # We can make this less generic when we stick with a single test pattern in
     # the repo.
     */conftest.py
     */test_*.py
     */*_tests.py
+    */test/*
 
 [report]
-fail_under = 25
+fail_under = 29
```
Comment (ashcrow) [inline .coveragerc]: :+1:

File: test-requirements.txt [hunk 11]
```diff
@@ -1,11 +1,9 @@
-six
-pyOpenSSL
+# flake8 must be listed before pylint to avoid dependency conflicts
```
Comment (tbielawa) [inline test-requirements.txt]: :+1:

File: .travis.yml [hunk 12]
```diff
@@ -4,17 +4,34 @@ sudo: false
...
+  email:
+    recipients:
+      - jdetiber@redhat.com
+      - sdodson@redhat.com
```
Comment (ashcrow) [inline .travis.yml]: This might be very noisy. I just want to make sure @sdodson is :+1: for that. I'd hate for the emails to be to much and simply be deleted.

**Comment by detiber** [comment]
aos-ci-test

File: .travis.yml (hunk 12 above)
Comment (detiber) [inline .travis.yml]: @ashcrow so, these notifications are for when the branch is broken post-commit. Mainly needed because of the issues with generated modules being out of sync in master due to conflicting PRs getting merged (where they individually passed generate_validation testing).

File: .coveragerc (hunk 10 above)
Comment (detiber) [inline .coveragerc]: @ashcrow I wouldn't get too excited, the coverage requirement for the separate utils tests was 73% :)

**Comment by detiber** [comment]
@tbielawa latest commit removes the Makefile and updates the testing docs.

**Comment by detiber** [comment]
Merging, since these changes only affect travis and only changes after approvals were to update docs and remove the no longer used Makefile.

File: .travis.yml (hunk 12 above)
Comment (rhcarvalho) [inline .travis.yml]: This shouldn't happen... we'd need a merge queue to prevent that, right?

File: utils/Makefile [hunk 13]
```diff
@@ -1,105 +0,0 @@
...
-clean:
-	@find . -type f -regex ".*\.py[co]$$" -delete
-	@find . -type f \( -name "*~" -or -name "#*" \) -delete
-	@rm -fR build dist rpm-build MANIFEST htmlcov .coverage cover ooinstall.egg-info oo-install
-	@rm -fR $(VENV)
-	@rm -fR .tox
-
-# To force a rebuild of the docs run 'touch' on any *.in file under
-# docs/man/man1/
-docs: $(MANPAGES)
```
Comment (rhcarvalho) [inline utils/Makefile]: The testing and linting part I know we had tox; this docs, man pages, etc was not used?

File: .travis.yml (hunk 12 above)
Comment (detiber) [inline .travis.yml]: We have a few other options to prevent it, but they all have drawbacks.

We could force PRs to be up to date with HEAD prior to merge, but that would cause everyone a lot of pain and frustration.

Forcing rebase merges might help as well, but I worry it would cause other potential issues.

With the CI/CD work, I think leveraging that for a merge queue (when ready) is the least painful approach.

File: utils/Makefile (hunk 13 above)
Comment (tbielawa) [inline utils/Makefile]: I think support for the quick installer is no longer a top priority. We had `$(MANPAGES)` there just to generate the one man page for `atomic-openshift-installer`. I don't see us needing to update that much more...
